src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -142,14 +181,25 @@ public synchronized void add(DecoratedKey key,
 
                 try
                 {
-                    data.putSingleton(encodedTerm, primaryKey, primaryKeysReducer, term.limit() <= MAX_RECURSIVE_KEY_LENGTH);
+                    data.putSingleton(encodedTerm, primaryKey, (existing, update) -> {
+                        // First do the normal primary keys reduction
+                        PrimaryKeys result = primaryKeysReducer.apply(existing, update);
+
+                        // Then update term frequency
+                        var pkbc = new PkWithTerm(update, encodedTerm);
+                        termFrequencies.merge(pkbc, 1, Integer::sum);
+
+                        return result;
+                    }, term.limit() <= MAX_RECURSIVE_KEY_LENGTH);
```

Comment:
One minor issue that could lead to surprises is that overwritten rows in the same memtable will lead to terms with incorrect frequencies but correct `docLengths`. This is because `IndexContext::update` is not properly implemented for non-vector indexes. It's an edge case, but probably one that could be hit in early manual testing.

I started on the implementation to support updates here https://github.com/datastax/cassandra/pull/1200, but priorities shifted and it was not fully reviewed. It seems like it would be a good idea to complete that work once this is merged so that we can have correct frequencies.